### PR TITLE
feat(gui): up to 6 additional domains with 256 char length

### PR DESCRIPTION
**Pull request by nikolagigic**
Ticket: #23544

### CODE REVIEW

File: gui/network/forms.py [hunk 1]
```diff
@@ -644,9 +644,13 @@ def clean_gc_netwait_ip(self):
 
     def clean(self):
         cleaned_data = self.cleaned_data
+        domains = cleaned_data.get("gc_domains")
         nameserver1 = cleaned_data.get("gc_nameserver1")
         nameserver2 = cleaned_data.get("gc_nameserver2")
         nameserver3 = cleaned_data.get("gc_nameserver3")
+        if len(domains.split()) > 6:
```
Comment (william-gr) [inline gui/network/forms.py]: You probably need to process that field so the entries are properly added to the database, without extra spaces and/or newlines:
`if domains:
    domains = domains.split()
    cleaned_data['gc_domains'] = '\n'.join(domains)
    if len(domains) > 5:
```

Not the `> 5`, you can have at max 6, you already have one from the main domain.

File: gui/network/forms.py [hunk 2]
```diff
@@ -644,9 +644,13 @@ def clean_gc_netwait_ip(self):
...
+            msg = _("More than 6 domain names.")
```
Comment (william-gr) [inline gui/network/forms.py]: "No more than 5 additional domains are allowed."

reads a little better.

File: gui/network/models.py [hunk 3]
```diff
@@ -70,6 +70,13 @@ class GlobalConfiguration(Model):
             regex=r'^[a-zA-Z\.\-\_0-9]+$',
         )],
     )
+    gc_domains = models.TextField(
+        max_length=256,
+        verbose_name=_("Additional domains"),
+        default='',
+        blank=True,
+        help_text="Delimit domains by spaces."
```
Comment (william-gr) [inline gui/network/models.py]: You're missing a translation of help_text, using _("")

Also, that text is not very descriptive. @wblock can you please help us here with something better to write here? (refer to man resolv.conf, "search" directive).

File: gui/network/migrations/0006_globalconfiguration_gc_domains.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
+# -*- coding: utf-8 -*-
+# Generated by Django 1.10.7 on 2017-08-23 11:07
+from __future__ import unicode_literals
+
+from django.db import migrations, models
+
+
+class Migration(migrations.Migration):
+    dependencies = [
+        ('network', '0005_vlan_pcp.py')
```
Comment (william-gr) [inline gui/network/migrations/0006_globalconfiguration_gc_domains.py]: Did you manually fiddle with that like, it should not have ended with .py, this is leading to the error:
django.db.migrations.exceptions.NodeNotFoundError: Migration network.0006_globalconfiguration_gc_domains dependencies reference nonexistent parent node ('network', '0005_vlan_pcp.py')

Comment (nikolagigic) [inline gui/network/migrations/0006_globalconfiguration_gc_domains.py]: Yes I wrote the migration script manually and accidently added .py in the end. 
I will remove it

File: src/middlewared/middlewared/plugins/network.py [hunk 5]
```diff
@@ -480,7 +480,9 @@ class DNSService(Service):
         else:
             gc = await self.middleware.call('datastore.query', 'network.globalconfiguration', None, {'get': True})
             if gc['gc_domain']:
-                domain = gc['gc_domain']
+                domains.append('{}\n'.format(gc['gc_domain']))
+            if gc['gc_domains']:
+                domains.append(gc['gc_domains'])
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/network.py]: if you do that you end up with a domains list like: ['domain1', 'extra1\nextra2\nextra3']

You would want ['domain1', 'extra1', 'extra2', 'extra3']

File: src/middlewared/middlewared/plugins/network.py [hunk 6]
```diff
@@ -489,8 +491,11 @@ class DNSService(Service):
                 nameservers.append(gc['gc_nameserver3'])
 
         resolvconf = ''
-        if domain:
-            resolvconf += 'search {}\n'.format(domain)
+
+        if domains:
+            for domain in domains:
+                resolvconf += 'search {}'.format(domain)
+
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/network.py]: You will end up with several seach entries, which seems wrong.

You want "search" followed by all domains, space delimited.

File: src/middlewared/middlewared/plugins/network.py [hunk 7]
```diff
@@ -480,7 +480,10 @@ class DNSService(Service):
         else:
             gc = await self.middleware.call('datastore.query', 'network.globalconfiguration', None, {'get': True})
             if gc['gc_domain']:
-                domain = gc['gc_domain']
+                domains.append('{}\n'.format(gc['gc_domain']))
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/network.py]: I think you misunderstood me here.

Each item of the domain list, must be a domain, no spaces or new lines.

File: src/middlewared/middlewared/plugins/network.py [hunk 8]
```diff
@@ -480,7 +480,10 @@ class DNSService(Service):
...
+            if gc['gc_domains']:
+                for domain in gc['gc_domains'].split():
+                    domains.append(domain)
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/network.py]: domains += gc['gc_domains'].split()

Is easier ;)

File: src/middlewared/middlewared/plugins/network.py [hunk 9]
```diff
@@ -489,8 +492,13 @@ class DNSService(Service):
...
         resolvconf = ''
-        if domain:
-            resolvconf += 'search {}\n'.format(domain)
+
+        if domains:
+            resolvconf += "search"
+            for domain in domains:
+                resolvconf += ' {}'.format(domain)
+            resolvconf += '\n'
+
```
Comment (william-gr) [inline src/middlewared/middlewared/plugins/network.py]: that does not look quite right, try this:
```
      if domain:
          resolvconf += 'search {}\n'.format(' '.join(domains))
```

**Comment by kmoore134** [comment]
Jenkins: retest this please

**Comment by william-gr** [comment]
Jenkins: retest this please

File: gui/network/forms.py [hunk 10]
```diff
@@ -644,9 +644,17 @@ def clean_gc_netwait_ip(self):
 
     def clean(self):
         cleaned_data = self.cleaned_data
+        domains = cleaned_data.get("gc_domains")
         nameserver1 = cleaned_data.get("gc_nameserver1")
         nameserver2 = cleaned_data.get("gc_nameserver2")
         nameserver3 = cleaned_data.get("gc_nameserver3")
+        if domains:
+            if len(domains) > 5:
+                msg = _("No more than 5 additional domains are allowed.")
```
Comment (wblock) [inline gui/network/forms.py]: Up to 5 additional domains are allowed.

File: gui/network/forms.py [hunk 11]
```diff
@@ -644,9 +644,17 @@ def clean_gc_netwait_ip(self):
...
+                self._errors["gc_domains"] = self.error_class([msg])
+            else:
+                domains = domains.split()
+                cleaned_data['gc_domains'] = '\n'.join(domains)
         if nameserver3:
```
Comment (wblock) [inline gui/network/forms.py]: All of these static checks are redundant.  Why not a loop?

File: gui/network/migrations/0006_globalconfiguration_gc_domains.py [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+    operations = [
+        migrations.AddField(
+            model_name='GlobalConfiguration',
+            name='gc_domains',
+            field=models.TextField(
+                max_length=256,
+                verbose_name="Additional domains",
+                default='',
+                blank=True,
+                help_text="Delimit domains by spaces.")
```
Comment (wblock) [inline gui/network/migrations/0006_globalconfiguration_gc_domains.py]: "Domain names must be separated by spaces."

File: gui/network/models.py [hunk 13]
```diff
@@ -70,6 +70,13 @@ class GlobalConfiguration(Model):
             regex=r'^[a-zA-Z\.\-\_0-9]+$',
         )],
     )
+    gc_domains = models.TextField(
+        max_length=256,
+        verbose_name=_("Additional domains"),
+        default='',
+        blank=True,
+        help_text=_("Delimit domains by spaces.")
```
Comment (wblock) [inline gui/network/models.py]: "Domain names must be separated by spaces."

**Pull request closed without merging**